packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -0,0 +1,254 @@
+    uint256 sellAmount;
+    // The amount of the buy token being bought. For stable tokens, this is
+    // kept track of as the value, not units.
+    uint256 buyAmount;
+    // The timestamp (`block.timestamp`) at which the exchange proposal was approved.
+    // If the exchange proposal has not ever been approved, is 0.
+    uint256 approvalTimestamp;
+    // The state of the exchange proposal.
+    ExchangeState state;
+    // Whether CELO is being sold and stableToken is being bought.
+    bool sellCelo;
+  }
+
+  // The percent fee imposed upon an exchange execution.
+  FixidityLib.Fraction public spread;
+
+  // The minimum and maximum amount of the stable token that can be minted or
+  // burned in a single exchange. Indexed by stable token address.
+  mapping(address => ExchangeLimits) public stableTokenExchangeLimits;
+
+  // State for all exchange proposals. Indexed by the exchange proposal ID.
+  mapping(uint256 => ExchangeProposal) public exchangeProposals;
+
+  // Number of exchange proposals that exist. Used for assigning an exchange
+  // proposal ID to a new proposal.
+  uint256 public exchangeProposalCount;
+
+  /**
+   * @notice Sets initialized == true on implementation contracts.
+   * @param test Set to true to skip implementation initialization.
+   */
+  constructor(bool test) public InitializableV2(test) {}
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _registry The address of the registry.
+   * @param _spread The spread charged on exchanges.
+   */
+  function initialize(address _registry, uint256 _spread) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(_registry);
+    setSpread(_spread);
+  }
+
+  /**
+   * @notice Creates a new exchange proposal and deposits the tokens being sold.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The proposal identifier for the newly created exchange proposal.
+   */
+  function createExchangeProposal(address stableToken, uint256 sellAmount, bool sellCelo)
+    external
+    nonReentrant
+    returns (uint256)
+  {
+    // Require the configurable stableToken max exchange amount to be > 0.
+    // This covers the case where a stableToken has never been explicitly permitted.
+    ExchangeLimits memory exchangeLimits = stableTokenExchangeLimits[stableToken];
+    require(exchangeLimits.maxExchangeAmount > 0, "Max stable token exchange amount must be > 0");
+
```

Comment:
Should you also check it's initialized?

Reply:
Following what's also been done for other contracts, I think it's safe to not explicitly check if it's been initialized. It's implied that the contract has been initialized because the initialization is done at the same time as setting the implementation to the proxy in [`_setAndInitializeImplementation`](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/common/Proxy.sol#L101)